README.md
[![Build Status](https://travis-ci.org/binary-com/perl-pubnub-pubsub.svg?branch=master)](https://travis-ci.org/binary-com/perl-pubnub-pubsub)
[![codecov](https://codecov.io/gh/binary-com/perl-pubnub-pubsub/branch/master/graph/badge.svg)](https://codecov.io/gh/binary-com/perl-pubnub-pubsub)
[![Gitter chat](https://badges.gitter.im/binary-com/perl-pubnub-pubsub.png)](https://gitter.im/binary-com/perl-pubnub-pubsub)

# NAME
[![Gitter](https://badges.gitter.im/Join Chat.svg)](https://gitter.im/binary-com/perl-pubnub-pubsub?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

PubNub::PubSub - Perl library for rapid publishing of messages on PubNub.com

# SYNOPSIS

    use PubNub::PubSub;

    my $pubnub = PubNub::PubSub->new(
        pub_key => 'demo', # only required for publish
        sub_key => 'demo',
        channel => 'sandbox',
    );

    # publish
    $pubnub->publish({
        messages => ['message1', 'message2'],
        callback => sub {
            my ($res) = @_;

            # $res is a L<Mojo::Message::Response>
            say $res->code; # 200
            say Dumper(\$res->json); # [1,"Sent","14108733777591385"]
        }
    });
    $pubnub->publish({
        channel  => 'sandbox2', # optional, if not applied, the one in ->new will be used.
        messages => ['message3', 'message4']
    });

    # subscribe
    $pubnub->subscribe({
        callback => sub {
            my (@messages) = @_;
            foreach my $msg (@messages) {
                print "# Got message: $msg\n";
            }
            return 1; # 1 to continue, 0 to stop
        }
    });

# DESCRIPTION

PubNub::PubSub is Perl library for rapid publishing of messages on PubNub.com based on [Mojo::UserAgent](https://metacpan.org/pod/Mojo::UserAgent)

perl clone of [https://gist.github.com/stephenlb/9496723#pubnub-http-pipelining](https://gist.github.com/stephenlb/9496723#pubnub-http-pipelining)

For a rough test:

- run perl examples/subscribe.pl in one terminal (or luanch may terminals with subscribe.pl)
- run perl examples/publish.pl in another terminal (you'll see all subscribe terminals will get messages.)

# METHOD

## new

- pub\_key

    optional, default pub\_key for publish

- sub\_key

    optional, default sub\_key for all methods

- channel

    optional, default channel for all methods

- publish\_callback

    optional. default callback for publish

- debug

    set ENV MOJO\_USERAGENT\_DEBUG to debug

## subscribe

subscribe channel to listen for the messages.

    $pubnub->subscribe({
        callback => sub {
            my (@messages) = @_;
            foreach my $msg (@messages) {
                print "# Got message: $msg\n";
            }
            return 1; # 1 to continue, 0 to stop
        }
    });

return 0 to stop

## publish

publish messages to channel

    $pubnub->publish({
        messages => ['message1', 'message2'],
        callback => sub {
            my ($res) = @_;

            # $res is a L<Mojo::Message::Response>
            say $res->code; # 200
            say Dumper(\$res->json); # [1,"Sent","14108733777591385"]
        }
    });
    $pubnub->publish({
        channel  => 'sandbox2', # optional, if not applied, the one in ->new will be used.
        messages => ['message3', 'message4']
    });

Note if you need shared callback, please pass it when do ->new with **publish\_callback**.

new Parameters specifically for **Publish V2 ONLY**

- ortt - Origination TimeToken where "r" = DOMAIN and "t" = TIMETOKEN
- meta - any JSON payload - intended as a safe and unencrypted payload
- ear - Eat At Read (read once)
- seqn - Sequence Number - for Guaranteed Delivery/Ordering

We'll first try to read from **messages**, if not specified, fall back to the same level as messages. eg:

    $pubnub->publish({
        messages => [
            {
                message => 'test message.',
                ortt => {
                    "r" => 13,
                    "t" => "13978641831137500"
                },
                meta => {
                    "stuff" => []
                },
                ear  => 'True',
                seqn => 12345,
            },
            {
                ...
            }
        ]
    });

    ## if you have common part, you can specified as the same level as messages
    $pubnub->publish({
        messages => [
            {
                message => 'test message.',
                ortt => {
                    "r" => 13,
                    "t" => "13978641831137500"
                },
                seqn => 12345,
            },
            {
                ...
            }
        ],
        meta => {
            "stuff" => []
        },
        ear  => 'True',
    });

## history

fetches historical messages of a channel

- sub\_key

    optional, default will use the one passed to ->new

- channel

    optional, default will use the one passed to ->new

- count

    Specifies the number of historical messages to return. The Default is 100.

- reverse

    Setting to true will traverse the time line in reverse starting with the newest message first. Default is false. If both start and end arguments are provided, reverse is ignored and messages are returned starting with the newest message.

- start

    Time token delimiting the start of time slice (exclusive) to pull messages from.

- end

    Time token delimiting the end of time slice (inclusive) to pull messages from.

Sample code:

    my $history = $pubnub->history({
        count => 20,
        reverse => "false"
    });
    # $history is [["message1", "message2", ... ],"Start Time Token","End Time Token"]

for example, to fetch all the rows in history

    my $history = $pubnub->history({
        reverse => "true",
    });
    while (1) {
        print Dumper(\$history);
        last unless @{$history->[0]}; # no messages
        sleep 1;
        $history = $pubnub->history({
            reverse => "true",
            start => $history->[2]
        });
    }

# GITHUB

[https://github.com/binary-com/perl-pubnub-pubsub](https://github.com/binary-com/perl-pubnub-pubsub)

# AUTHOR

Binary.com <fayland@gmail.com>

# LICENSE AND COPYRIGHT

Copyright 2014- binary.com.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

[http://www.perlfoundation.org/artistic\_license\_2\_0](http://www.perlfoundation.org/artistic_license_2_0)

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Changes
Revision history for PubNub-PubSub

{{$NEXT}}

1.0.2  2019-01-05 23:41:19+00:00 UTC
       - use Mojo::UserAgent::CookieJar behavior from Mojolicious >= 6.20 and bump dependency
       - add simple testcase for subscribe method

1.0.1  2017-02-20 06:36:08+00:00 UTC
       - switch to dzil
  
1.0.0

0.08  2014-11-01 12:18:59 CST
        - add new params for Publish V2

0.07  2014-09-16 21:24:39 CST
        - all use LWP::UserAgent, publish callback API changes.

0.06  2014-09-15 17:34:06 CST
        - rewrite history to support v2 http://www.pubnub.com/docs/javascript/api/reference.html#history

0.05  2014-09-11 16:02:08 CST
        - more elegant methods to call (API Changes, I bet you haven't use the latest version in such short time.)

0.04  2014-09-11 14:55:35 CST
        - API Changes. publish only accepts messages and all others should be passed into ->new
        - use one connection for all publish no matter how many times it's called

0.03  2014-09-09 16:57:12 CST
        - fix the some old perl? pass more args to the __r sub

0.02  2014-09-05 19:16:43 CST
        - for old Mojo version

0.01  2014-09-04 09:38:13 CST
        - Initial release


examples/test.pl
#!/usr/bin/perl

use strict;
use warnings;
use FindBin qw/$Bin/;
use lib "$Bin/../lib";
use PubNub::PubSub;
use Time::HiRes qw/time/;

srand();
my $total_message = 50000 + int(rand(50000));

#  s.send('GET /publish/pub-c-5afaf11d-aa91-4a40-b0d2-77961fb3a258/sub-c-0cd3a376-28ac-11e4-95a7-02ee2ddab7fe/0/HyperLogLogDemo1/0/"'+str(id[random.randrange(0, numID-1)])+'" HTTP/1.1\r\nHost: pubsub.pubnub.com\r\n\r\n')
my $got_message = 0;
my $start_time = time();
my $pubnub = PubNub::PubSub->new(
    pub_key => 'pub-c-5afaf11d-aa91-4a40-b0d2-77961fb3a258',
    sub_key => 'sub-c-0cd3a376-28ac-11e4-95a7-02ee2ddab7fe',
    channel => 'HyperLogLogDemo1',
    callback => sub {
        my ($res, $req) = @_;

        # print "=" x 20 . "\n";
        # print "REQ: $req" . "\n";
        # print "=" x 20 . "\n";
        # print "RES: $res" . "\n";
        # print "=" x 20 . "\n";

        $got_message++;
        if ($got_message == $total_message or $got_message % 1000 == 0) {
            my $duration = time() - $start_time;
            print "$got_message Spent $duration.\n";
        }
    }
);

print "Total sending $total_message\n";
my @messages;
foreach (1 .. $total_message) {
    push @messages, "message" . int(rand(100000));
}
$pubnub->publish(@messages);

print "Total got $got_message VS $total_message\n";
my $duration = time() - $start_time;
print "Spent $duration.\n";

1;

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.007.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker 6.48;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl library for rapid publishing of messages on PubNub.com",
  "AUTHOR" => "binary.com <BINARY\@cpan.org>",
  "BUILD_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.48"
  },
  "DISTNAME" => "PubNub-PubSub",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "PubNub::PubSub",
  "PREREQ_PM" => {
    "Mojolicious" => "6.20"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::More" => "0.96"
  },
  "VERSION" => "1.0.1",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Mojolicious" => "6.20",
  "Test::CheckDeps" => "0.010",
  "Test::More" => "0.96"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


examples/publish.pl
#!/usr/bin/perl

use strict;
use warnings;
use FindBin qw/$Bin/;
use lib "$Bin/../lib";
use PubNub::PubSub;
use Time::HiRes qw/time/;
use Data::Dumper;

srand();
my $total_message = 10; # 50000 + int(rand(50000));
my $got_message = 0;
my $start_time = time();

my $pubnub = PubNub::PubSub->new(
    pub_key  => $ENV{PUBNUB_PUB_KEY} || 'pub-c-5b5d836f-143b-48d2-882f-659e87b6c321',
    sub_key  => $ENV{PUBNUB_SUB_KEY} || 'sub-c-a66b65f2-2d96-11e4-875c-02ee2ddab7fe',
    channel  => $ENV{PUBNUB_CHANNEL} || 'sandbox',
);

print "Total sending $total_message\n";
my @messages;
foreach (1 .. $total_message) {
    push @messages, "message" . $_;
}
$pubnub->publish({
    messages => \@messages,
    callback => sub {
        my ($res) = @_;

        print Dumper(\$res->json);

        $got_message++;
        if ($got_message == $total_message or $got_message % 1000 == 0) {
            my $duration = time() - $start_time;
            print "$got_message spent $duration.\n";
        }
    }
});

my $duration = time() - $start_time;
print "Spent $duration.\n";

1;

examples/publish_multi.pl
#!/usr/bin/perl

use strict;
use warnings;
use FindBin qw/$Bin/;
use lib "$Bin/../lib";
use PubNub::PubSub;
use Time::HiRes qw/time/;
use Data::Dumper;

srand();
my $total_message = 10; # 50000 + int(rand(50000));

my $got_message = 0;
my $start_time = time();
my $pubnub = PubNub::PubSub->new(
    pub_key  => $ENV{PUBNUB_PUB_KEY} || 'pub-c-5b5d836f-143b-48d2-882f-659e87b6c321',
    sub_key  => $ENV{PUBNUB_SUB_KEY} || 'sub-c-a66b65f2-2d96-11e4-875c-02ee2ddab7fe',
    channel  => $ENV{PUBNUB_CHANNEL} || 'sandbox',
    publish_callback => sub {
        my ($data) = @_;

        print Dumper(\$data);

        $got_message++;
        if ($got_message == $total_message or $got_message % 1000 == 0) {
            my $duration = time() - $start_time;
            print "$got_message Spent $duration.\n";
        }
    }
);

print "sending $total_message messages\n";
my @messages;
foreach (1 .. $total_message) {
    push @messages, "message" . $_;;
}
$pubnub->publish({ messages => \@messages });

print "sending another $total_message messages\n";
@messages = ();
foreach ($total_message + 1 .. $total_message * 2) {
    push @messages, "message" . $_;;
}
$pubnub->publish({ messages => \@messages });

1;

lib/PubNub/PubSub/Message.pm
package PubNub::PubSub::Message;

use Carp;
use Encode;
use JSON::MaybeXS;

use strict;
use warnings;

## VERSION

=head1 NAME

PubNub::PubSub::Message - Message object for PubNub::PubSub

=head1 SYNOPSIS

This module is primarily used behind the scenes in PubNub::PubSub.  It is
not intended to be used directly for users.  This being said, one can use it
if you want to do your own URL management or otherwise interface with PubNub in
ways this distribution does not yet support.

 my $datastructure;
 my $message = PubNub::PubSub::Message->new(payload=> $datastructure);
 my $json = $message->json;
 my $payload = $message->payload;
 my $queryhash = $message->query_params;

=head1 METHODS

=head2 new

THis is the basic constructor.  Requires message or payload argument.  Message
is effectively an alias for payload.  Other arguments include ortt, meta, ear,
and seqn, supported per the PubNub API.  These other arguments are converted
to JSON in the query_params method below.

If a simple scalar is passed (not a reference), it is assumed that this will
be passed to PubNub as a string literal and handled appropriately.

=cut


sub new { ## no critic (RequireArgUnpacking)
    my $pkg  = shift;
    unshift @_, 'payload' if scalar @_ == 1 && !ref $_[0];
    my %args = scalar @_ % 2 ? %{$_[0]} : @_;
    $args{payload} ||= $args{message}; # backwards compatibility
    croak 'Must provide payload' unless $args{payload};
    my $self = \%args;
    return bless $self, $pkg;
}

=head2 payload

Returns the message payload

=cut

sub payload {
    my $self = shift;
    return $self->{payload};
}

=head2 from_msg($json_string)

Returns a message object with a payload from a json string.

=cut
my $json = JSON::MaybeXS->new->allow_nonref(1);
sub from_msg {
    my ($self, $json) = @_;
    my $arrayref = $json->decode(Encode::decode_utf8($json));
    return "$self"->new(payload => $arrayref->[0], timestamp => $arrayref->[1]);
}

=head2 json

Returns the payload encoded in json

=cut

sub json {
    my $self = shift;
    return Encode::encode_utf8($json->encode($self->{payload}));
}

=head2 query_params($mergehash)

Returns a hash of query param properties (ortt, meta, ear, seqn), json-encoded,
for use in constructing URI's for PubNub requests.

=cut

sub query_params {
    my $self = shift;
    my $merge = shift;
    return { map {
        my $var = $self->{$_};
        $var = $merge->{$_} unless defined $var;
        defined $var ?
            ($_ => Encode::encode_utf8($json->encode($var))) :
            ();
    } qw(ortt meta ear seqn) };
}

1;


lib/PubNub/PubSub.pm
package PubNub::PubSub;

use strict;
use warnings;
use v5.10;

use Carp;
use Mojo::UserAgent;
use Mojo::Util qw/url_escape/;

use PubNub::PubSub::Message;

our $VERSION = '1.0.2';

sub new { ## no critic (RequireArgUnpacking)
    my $class = shift;
    my %args  = @_ % 2 ? %{$_[0]} : @_;

    $args{host} ||= 'pubsub.pubnub.com';
    $args{port} ||= 80;
    $args{timeout} ||= 60; # for ua timeout
    $args{publish_queue} ||= [];

    my $proto = ($args{port} == 443) ? 'https://' : 'http://';
    $args{web_host} ||= $proto . $args{host};

    return bless \%args, $class;
}

sub __ua {
    my $self = shift;

    return $self->{ua} if exists $self->{ua};

    my $ua = Mojo::UserAgent->new;
    $ua->max_redirects(3);
    $ua->inactivity_timeout($self->{timeout});
    $ua->proxy->detect; # env proxy
    $ua->cookie_jar->ignore(sub { 1 });
    $ua->max_connections(100);
    $self->{ua} = $ua;

    return $ua;
}

sub publish { ## no critic (RequireArgUnpacking)
    my $self = shift;

    my %params = @_ % 2 ? %{$_[0]} : @_;
    my $callback = $params{callback} || $self->{publish_callback};

    my $ua = $self->__ua;

    my @steps = map {
             my $ref = $_;
             my $url = $ref->{url};
             sub {
                 my $delay = shift;
                 my $end = $delay->begin;
                 $ua->get($url => sub {
                    $callback->($_[1]->res, $ref->{message}) if $callback;
                    $end->();
                  });
             }
    } $self->__construct_publish_urls(%params);

    return Mojo::IOLoop->delay(@steps)->wait;
}

sub __construct_publish_urls {
    my ($self, %params) = @_;

    my $pub_key = $params{pub_key} || $self->{pub_key};
    $pub_key or croak "pub_key is required.";
    my $sub_key = $params{sub_key} || $self->{sub_key};
    $sub_key or croak "sub_key is required.";
    my $channel = $params{channel} || $self->{channel};
    $channel or croak "channel is required.";
    $params{messages} or croak "messages is required.";

    return map {
        my $json = $_->json;
        my $uri = Mojo::URL->new( $self->{web_host} . qq~/publish/$pub_key/$sub_key/0/$channel/0/~ . url_escape($json) );
        $uri->query($_->query_params(\%params));
        { url => $uri->to_string, message => $_ };
    } map { PubNub::PubSub::Message->new($_) } @{$params{messages}};
}

sub subscribe { ## no critic (RequireArgUnpacking)
    my $self = shift;
    my %params = @_ % 2 ? %{$_[0]} : @_;

    my $sub_key = $params{sub_key} || $self->{sub_key};
    $sub_key or croak "sub_key is required.";
    my $channel = $params{channel} || $self->{channel};
    $channel or croak "channel is required.";

    my $callback = $params{callback} or croak "callback is required.";
    my $timetoken = $params{timetoken} || '0';

    my $ua = $self->__ua;

    my $tx = $ua->get($self->{web_host} . "/subscribe/$sub_key/$channel/0/$timetoken");
    unless ($tx->success) {
        # for example $tx->error->{message} =~ /Inactivity timeout/

        # This is not a traditional goto. Instead it exits this function 
        # and re-enters with @ as params.
        #
        # see goto docs, this is basically a method call which exits the current
        # function first.  So no extra call stack depth.
        sleep 1;
        @_ = ($self, %params, timetoken => $timetoken);
        goto &subscribe;
    }
    my $json = $tx->res->json;
    my @cb_args = $params{raw_msg}? ($json) : (@{$json->[0]});

    my $rtn = $callback ? $callback->(@cb_args) : 1;
    return unless $rtn;

    $timetoken = $json->[1];
    @_ = ($self, %params, timetoken => $timetoken);
    goto &subscribe;
}

sub subscribe_multi { ## no critic (RequireArgUnpacking)
    my $self = shift;
    my %params = @_ % 2 ? %{$_[0]} : @_;
    croak 'channels must be an arrayref'
         unless ref($params{channels}) =~ /ARRAY/;
    croak 'callback must be a hashref or coderef'
         unless ref($params{callback}) =~ /(HASH|CODE)/;

    my $callback;
    if (ref($params{callback}) =~ /HASH/){
       for (keys %{$params{callback}}) {
           croak "Non-coderef value found for callback key $_" 
                unless ref($params{callback}->{$_}) =~ /CODE/;
       }
       $callback = sub {
           my ($obj) = @_;
           my (undef, undef, $channel) = @$obj;
           my $cb_dispatch = $params{callback};
           unless ($channel) { # on connect messages
              goto $cb_dispatch->{on_connect}
                   if exists $cb_dispatch->{on_connect};
              return 1;
           }
           if (exists $cb_dispatch->{$channel}){

              # these are verified coderefs, so replacing the current stack 
              # frame with a call to the function.  They will *not* jump to 
              # a label or other points.  Basically this just lets us pretend
              # that this was called directly by subscribe above.
              goto $cb_dispatch->{$channel};
           } elsif (exists $cb_dispatch->{'_default'}){
              goto $cb_dispatch->{_default};
           } else {
              warn 'Using callback dispatch table, cannot find channel callback'
                   . ' and _default callback not specified';
              return;
           }
       };
    }
    $callback = $params{callback} unless ref $callback;

    my $channel_string = join ',', @{$params{channels}};
    return $self->subscribe(channel => $channel_string, callback => $callback,
                           raw_msg => 1);
}

sub history { ## no critic (RequireArgUnpacking)
    my $self = shift;

    if (scalar(@_) == 1 and ref($_[0]) ne 'HASH' and $_[0] =~ /^\d+$/) {
        @_ = (count => $_[0]);
        warn "->history(\$num) is deprecated and will be removed in next few releases.\n";
    }

    my %params = @_ % 2 ? %{$_[0]} : @_;

    my $sub_key = delete $params{sub_key} || $self->{sub_key};
    $sub_key or croak "sub_key is required.";
    my $channel = delete $params{channel} || $self->{channel};
    $channel or croak "channel is required.";

    my $ua = $self->__ua;

    my $tx = $ua->get($self->{web_host} . "/v2/history/sub-key/$sub_key/channel/$channel" => form => \%params);
    return [$tx->error->{message}] unless $tx->success;
    return $tx->res->json;
}

1;
__END__

=encoding utf-8

=head1 NAME

PubNub::PubSub - Perl library for rapid publishing of messages on PubNub.com

=head1 SYNOPSIS

    use PubNub::PubSub;
    use 5.010;
    use Data::Dumper;

    my $pubnub = PubNub::PubSub->new(
        pub_key => 'demo', # only required for publish
        sub_key => 'demo',
        channel => 'sandbox',
    );

    # publish
    $pubnub->publish({
        messages => ['message1', 'message2'],
        callback => sub {
            my ($res) = @_;

            # $res is a L<Mojo::Message::Response>
            say $res->code; # 200
            say Dumper(\$res->json); # [1,"Sent","14108733777591385"]
        }
    });
    $pubnub->publish({
        channel  => 'sandbox2', # optional, if not applied, the one in ->new will be used.
        messages => ['message3', 'message4']
    });

    # subscribe
    $pubnub->subscribe({
        callback => sub {
            my (@messages) = @_;
            foreach my $msg (@messages) {
                print "# Got message: $msg\n";
            }
            return 1; # 1 to continue, 0 to stop
        }
    });


=head1 DESCRIPTION

PubNub::PubSub is Perl library for rapid publishing of messages on PubNub.com based on L<Mojo::UserAgent>

perl clone of L<https://gist.github.com/stephenlb/9496723#pubnub-http-pipelining>

For a rough test:

=over 4

=item * run perl examples/subscribe.pl in one terminal (or luanch may terminals with subscribe.pl)

=item * run perl examples/publish.pl in another terminal (you'll see all subscribe terminals will get messages.)

=back

=head1 METHOD

=head2 new

=over 4

=item * pub_key

optional, default pub_key for publish

=item * sub_key

optional, default sub_key for all methods

=item * channel

optional, default channel for all methods

=item * publish_callback

optional. default callback for publish

=item * debug

set ENV MOJO_USERAGENT_DEBUG to debug

=back

=head2 subscribe

subscribe channel to listen for the messages.

Arguments are:

=over

=item callback

Callback to run on the channel

=item channel

Channel to listen on, defaults to the base object's channel attribute.

=item subkey

Subscription key.  Defaults to base object's subkey attribute.

=item raw_msg

Pass the whole message in, as opposed to the json element of the payload.

This is useful when you need to process time tokens or channel names.

The format is a triple of (\@messages, $timetoken, $channel).

=item timetoken

Time token for initial request.  Defaults to 0.

=back

    $pubnub->subscribe({
        callback => sub {
            my (@messages) = @_;
            foreach my $msg (@messages) {
                print "# Got message: $msg\n";
            }
            return 1; # 1 to continue, 0 to stop
        }
    });

return 0 to stop

=head2 subscribe_multi

Subscribe to multiple channels.  Arguments are:

=over

=item channels

an arrayref of channel names

=item callback

A callback, either a coderef which handles all requests, or a hashref dispatch
table with one entry per channel.

If a dispatch table is used a _default entry catches all unrecognized channels.
If an unrecognized channel is found, a warning is generated and the loop exits.

The message results are passed into the functions in raw_msg form (i.e. a tuple
ref of (\@messages, $timetoken, $channel) for performance reasons.

=back

=head2 publish

publish messages to channel

    $pubnub->publish({
        messages => ['message1', 'message2'],
        callback => sub {
            my ($res) = @_;

            # $res is a L<Mojo::Message::Response>
            say $res->code; # 200
            say Dumper(\$res->json); # [1,"Sent","14108733777591385"]
        }
    });
    $pubnub->publish({
        channel  => 'sandbox2', # optional, if not applied, the one in ->new will be used.
        messages => ['message3', 'message4']
    });

Note if you need shared callback, please pass it when do ->new with B<publish_callback>.

new Parameters specifically for B<Publish V2 ONLY>

=over 4

=item * ortt - Origination TimeToken where "r" = DOMAIN and "t" = TIMETOKEN

=item * meta - any JSON payload - intended as a safe and unencrypted payload

=item * ear - Eat At Read (read once)

=item * seqn - Sequence Number - for Guaranteed Delivery/Ordering

=back

We'll first try to read from B<messages>, if not specified, fall back to the same level as messages. eg:

    $pubnub->publish({
        messages => [
            {
                message => 'test message.',
                ortt => {
                    "r" => 13,
                    "t" => "13978641831137500"
                },
                meta => {
                    "stuff" => []
                },
                ear  => 'True',
                seqn => 12345,
            },
            {
                ...
            }
        ]
    });

    ## if you have common part, you can specified as the same level as messages
    $pubnub->publish({
        messages => [
            {
                message => 'test message.',
                ortt => {
                    "r" => 13,
                    "t" => "13978641831137500"
                },
                seqn => 12345,
            },
            {
                ...
            }
        ],
        meta => {
            "stuff" => []
        },
        ear  => 'True',
    });

=head2 history

fetches historical messages of a channel

=over 4

=item * sub_key

optional, default will use the one passed to ->new

=item * channel

optional, default will use the one passed to ->new

=item * count

Specifies the number of historical messages to return. The Default is 100.

=item * reverse

Setting to true will traverse the time line in reverse starting with the newest message first. Default is false. If both start and end arguments are provided, reverse is ignored and messages are returned starting with the newest message.

=item * start

Time token delimiting the start of time slice (exclusive) to pull messages from.

=item * end

Time token delimiting the end of time slice (inclusive) to pull messages from.

=back

Sample code:

    my $history = $pubnub->history({
        count => 20,
        reverse => "false"
    });
    # $history is [["message1", "message2", ... ],"Start Time Token","End Time Token"]

for example, to fetch all the rows in history

    my $history = $pubnub->history({
        reverse => "true",
    });
    while (1) {
        print Dumper(\$history);
        last unless @{$history->[0]}; # no messages
        sleep 1;
        $history = $pubnub->history({
            reverse => "true",
            start => $history->[2]
        });
    }

=head1 JSON USAGE

This module effectively runs a Mojolicious application in the background.  For
those parts of JSON which do not have a hard Perl equivalent, such as booleans,
the Mojo::JSON module's semantics work.  This means that JSON bools are
handled as references to scalar values 0 and 1 (i.e. \0 for false and \1 for
true).

This has changed since 0.08, where True and False were used.

=head1 GITHUB

L<https://github.com/binary-com/perl-pubnub-pubsub>

=head1 AUTHOR

Binary.com E<lt>fayland@gmail.comE<gt>

=cut
